flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java

```diff
@@ -270,10 +270,12 @@ private void updateLastSeenWatermark(long timestamp) {
 	}
 
 	private void updateNFA(NFA<IN> nfa) throws IOException {
-		if (nfa.isEmpty()) {
-			nfaOperatorState.clear();
-		} else {
-			nfaOperatorState.update(nfa);
+		if (nfa.isNFAChanged()) {
+			if (nfa.isEmpty()) {
+				nfaOperatorState.clear();
+			} else {
+				nfaOperatorState.update(nfa);
```

Comment:
Hmm, right we do access it from `StateBackend` each time. Not sure though if each time it is deserialized. I don't think there is such assumption. @kl0u could you shed some more light on it?

If I am right and it is not deserialized each time, I would reset the `nfaChanged` before writing it into the `StateBackend`.